File: scripts/create_icons.py
"""Generate extension icons using pure Python (zlib & struct)."""

import os
import struct
import zlib


def create_png(width: int, height: int, color: tuple[int, int, int, int]) -> bytes:
    """Create a minimal PNG file in bytes."""
    r, g, b, a = color
    line = bytes([0]) + bytes([r, g, b, a]) * width
    raw_data = line * height

    # PNG chunks
    ihdr_data = struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0)
    ihdr_crc = zlib.crc32(b"IHDR" + ihdr_data)
    ihdr_chunk = struct.pack(">I", len(ihdr_data)) + b"IHDR" + ihdr_data + struct.pack(">I", ihdr_crc)

    idat_compressed = zlib.compress(raw_data)
    idat_crc = zlib.crc32(b"IDAT" + idat_compressed)
    idat_chunk = struct.pack(">I", len(idat_compressed)) + b"IDAT" + idat_compressed + struct.pack(">I", idat_crc)

    iend_crc = zlib.crc32(b"IEND")
    iend_chunk = struct.pack(">I", 0) + b"IEND" + struct.pack(">I", iend_crc)

    png_header = b"\x89PNG\r\n\x1a\n"
    return png_header + ihdr_chunk + idat_chunk + iend_chunk


def generate_all_icons():
    icons_dir = os.path.join("extension", "icons")
    os.makedirs(icons_dir, exist_ok=True)

    sizes = [16, 32, 48, 128]
    # Indigo color: #6366f1 (99, 102, 241, 255)
    color = (99, 102, 241, 255)

    for size in sizes:
        filename = os.path.join(icons_dir, f"icon{size}.png")
        png_data = create_png(size, size, color)
        with open(filename, "wb") as f:
            f.write(png_data)
        print(f"Created {filename}")


if __name__ == "__main__":
    generate_all_icons()

File: scripts/create_icons.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

function createPng(width, height, r, g, b, a) {
    const rawData = [];
    for (let y = 0; y < height; y++) {
        rawData.push(0); // filter byte
        for (let x = 0; x < width; x++) {
            rawData.push(r, g, b, a);
        }
    }
    const rawBuffer = Buffer.from(rawData);

    // PNG header
    const signature = Buffer.from([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]);

    // IHDR
    const ihdr = Buffer.alloc(13);
    ihdr.writeUInt32BE(width, 0);
    ihdr.writeUInt32BE(height, 4);
    ihdr[8] = 8; // bit depth
    ihdr[9] = 6; // color type RGBA
    ihdr[10] = 0; // compression
    ihdr[11] = 0; // filter
    ihdr[12] = 0; // interlace
    const ihdrChunk = createChunk('IHDR', ihdr);

    // IDAT
    const compressed = zlib.deflateSync(rawBuffer);
    const idatChunk = createChunk('IDAT', compressed);

    // IEND
    const iendChunk = createChunk('IEND', Buffer.alloc(0));

    return Buffer.concat([signature, ihdrChunk, idatChunk, iendChunk]);
}

function createChunk(type, data) {
    const len = data.length;
    const buf = Buffer.alloc(8 + len + 4);
    buf.writeUInt32BE(len, 0);
    buf.write(type, 4);
    data.copy(buf, 8);

    // Calculate CRC
    const crcBuf = buf.subarray(4, 8 + len);
    const crc = crc32(crcBuf);
    buf.writeUInt32BE(crc, 8 + len);
    return buf;
}

function crc32(buf) {
    let crc = -1;
    for (let i = 0; i < buf.length; i++) {
        let byte = buf[i];
        crc ^= byte;
        for (let j = 0; j < 8; j++) {
            crc = (crc >>> 1) ^ (-(crc & 1) & 0xedb88320);
        }
    }
    return (crc ^ -1) >>> 0;
}

const iconsDir = path.join(__dirname, '..', 'extension', 'icons');
if (!fs.existsSync(iconsDir)) {
    fs.mkdirSync(iconsDir, { recursive: true });
}

[16, 32, 48, 128].forEach(size => {
    const png = createPng(size, size, 99, 102, 241, 255);
    const file = path.join(iconsDir, `icon${size}.png`);
    fs.writeFileSync(file, png);
    console.log(`Created ${file}`);
});
